### PR TITLE
test(pageserver): add delta records tests for gc-compaction

### PR DESCRIPTION
**Pull request by skyzh**
## Problem

Part of https://github.com/neondatabase/neon/issues/8002

This pull request adds tests for bottom-most gc-compaction with delta records. Also fixed a bug in the compaction process that creates overlapping delta layers by force splitting at the original delta layer boundary.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [x] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/tenant/timeline/compaction.rs [hunk 1]
```diff
@@ -1043,7 +1047,15 @@ impl Timeline {
             // We have a list of deltas/images. We want to create image layers while collect garbages.
             for (key, lsn, val) in accumulated_values.iter().rev() {
                 if *lsn > horizon {
-                    keys_above_horizon.push((*key, *lsn, val.clone())); // TODO: ensure one LSN corresponds to either delta or image instead of both
+                    if let Some((_, prev_lsn, _)) = keys_above_horizon.last_mut() {
+                        if *prev_lsn == *lsn {
+                            // The case that we have an LSN with both data from the delta layer and the image layer. As
+                            // `ValueWrapper` ensures that an image is ordered before a delta at the same LSN, we simply
+                            // drop this delta and keep the image.
+                            continue;
```
Comment (problame) [inline pageserver/src/tenant/timeline/compaction.rs:1055]: Does this happen in real life or just in your test cases?

File: pageserver/src/tenant/timeline/compaction.rs [hunk 2]
```diff
@@ -1087,40 +1133,63 @@ impl Timeline {
...
+                &mut delta_values,
+                last_key,
+                &delta_split_points,
+                &mut current_delta_split_point,
+                self,
+                gc_cutoff,
+                ctx,
+            )
+            .await?,
+        );
```
Comment (problame) [inline pageserver/src/tenant/timeline/compaction.rs:1183]: Would be nice to restructure the loop code so we don't have to repeat the loop body right after it. What's stopping us? (No need to do it in this PR, but, would be a nice cleanup)

File: pageserver/src/tenant/timeline/compaction.rs [hunk 3]
```diff
@@ -1068,15 +1080,49 @@ impl Timeline {
...
+            last_key: Key,
+            delta_split_points: &[Key],
+            current_delta_split_point: &mut usize,
+            tline: &Arc<Timeline>,
+            gc_cutoff: Lsn,
+            ctx: &RequestContext,
+        ) -> anyhow::Result<Option<ResidentLayer>> {
+            // Check if we need to split the delta layer. We split at the original delta layer boundary to avoid
+            // overlapping layers.
+            let mut need_split = false;
```
Comment (problame) [inline pageserver/src/tenant/timeline/compaction.rs:1094]: Hm, I don't see immediately how that overlap can happen. ASCII art or a _small_ test case that fails before this change can help.

**Review by problame — APPROVED** [review]
Test changes LGTM

Were the changes in compaction.rs exposed by the added tests?

---

I don't _fully_ understand why the split point business is necessary. Happy to follow up on this post-merge though.

File: pageserver/src/tenant.rs [hunk 4]
```diff
@@ -6934,9 +6975,167 @@ mod tests {
...
+
+    #[tokio::test]
+    async fn test_simple_bottom_most_compaction_deltas() -> anyhow::Result<()> {
+        let harness = TenantHarness::create("test_simple_bottom_most_compaction_deltas")?;
+        let (tenant, ctx) = harness.load().await;
+
+        fn get_key(id: u32) -> Key {
+            // using aux key here b/c they are guaranteed to be inside `collect_keyspace`.
+            let mut key = Key::from_hex("620000000033333333444444445500000000").unwrap();
+            key.field6 = id;
```
[the remark this answers is not in the document]
Comment (skyzh) [reply pageserver/src/tenant.rs:6995]: Aux keyspace is already included so no need to add the keyspace

File: pageserver/src/tenant/timeline/compaction.rs (hunk 1 above)
Comment (skyzh) [reply pageserver/src/tenant/timeline/compaction.rs:1055]: This happens in real when we create an image layer at the end LSN of a delta layer.

**Comment by skyzh** [comment]
> Were the changes in compaction.rs exposed by the added tests?

Yes, the tests reveal some bugs within the compaction algorithm. When it comes to the k-merge re-implementation of that part, we can do a thorough review.

File: pageserver/src/tenant/timeline/compaction.rs (hunk 3 above)
Comment (skyzh) [reply pageserver/src/tenant/timeline/compaction.rs]: https://github.com/neondatabase/neon/pull/8078/commits/d3f7d5e8a7738cb035b62953cdd13bc1c7289bea

File: pageserver/src/tenant/timeline/compaction.rs (hunk 2 above)
Comment (skyzh) [reply pageserver/src/tenant/timeline/compaction.rs:1196]: added a TODO for now, there could be a way to make all of them into the loop.

File: pageserver/src/tenant/timeline/compaction.rs [hunk 5]
```diff
@@ -987,19 +989,28 @@ impl Timeline {
         };
         // Step 1: (In the future) construct a k-merge iterator over all layers. For now, simply collect all keys + LSNs.
         let mut all_key_values = Vec::new();
+        let mut delta_split_points = BTreeSet::new();
```
[the remark this answers is not in the document]
Comment (skyzh) [reply pageserver/src/tenant/timeline/compaction.rs:993]: fixed in https://github.com/neondatabase/neon/pull/8078/commits/d3f7d5e8a7738cb035b62953cdd13bc1c7289bea

File: pageserver/src/tenant.rs [hunk 6]
```diff
@@ -6732,49 +6732,49 @@ mod tests {
 
         // img layer at 0x10
         let img_layer = (0..10)
-            .map(|id| (get_key(id), test_img(&format!("value {id}@0x10"))))
+            .map(|id| (get_key(id), Bytes::from(format!("value {id}@0x10"))))
             .collect_vec();
 
         let delta1 = vec![
             // TODO: we should test a real delta record here, which requires us to add a variant of NeonWalRecord for testing purpose.
```
[the remark this answers is not in the document]
Comment (skyzh) [reply pageserver/src/tenant.rs]: fixed in https://github.com/neondatabase/neon/pull/8078/commits/d3f7d5e8a7738cb035b62953cdd13bc1c7289bea

File: pageserver/src/tenant.rs [hunk 7]
```diff
@@ -6934,9 +6975,167 @@ mod tests {
...
+                vec![delta1, delta2, delta3], // delta layers
+                vec![(Lsn(0x10), img_layer)], // image layers
+                Lsn(0x50),
+            )
+            .await?;
+        {
+            // Update GC info
+            let mut guard = tline.gc_info.write().unwrap();
+            guard.cutoffs.pitr = Lsn(0x30);
+            guard.cutoffs.horizon = Lsn(0x30);
```
[the remark this answers is not in the document]
Comment (skyzh) [reply pageserver/src/tenant.rs]: fixed in https://github.com/neondatabase/neon/pull/8078/commits/d3f7d5e8a7738cb035b62953cdd13bc1c7289bea

File: pageserver/src/tenant/timeline/compaction.rs (hunk 1 above)
Comment (problame) [reply pageserver/src/tenant/timeline/compaction.rs:1056]: Maybe the comment should mention that.

Comment (skyzh) [reply pageserver/src/tenant/timeline/compaction.rs]: added in https://github.com/neondatabase/neon/pull/8078/commits/b139c895755188ae7abc4852c88434ff91f83b43

**Pull request merged**